# 第五章围棋的功能

Go 的语法*tour de force*之一是通过其对高阶函数的支持，这在动态语言（如 Python 或 Ruby）中可以找到。正如我们将在本章中看到的，函数也是一个类型化实体，其值可以分配给变量。在本章中，我们将探讨 Go 中的功能，包括以下主题：

*   围棋功能
*   传递参数值
*   匿名函数和闭包
*   高阶函数
*   错误信号处理
*   推迟函数调用
*   功能恐慌与恢复

# 围棋功能

在 Go 中，函数是一流的类型化编程元素。声明的函数文本始终具有类型和值（定义的函数本身），并且可以选择性地绑定到命名标识符。因为函数可以用作数据，所以可以将它们分配给变量或作为其他函数的参数传递。

## 功能声明

在 Go 中声明函数采用下图所示的一般形式。此规范形式用于声明命名函数和匿名函数。

![Function declaration](graphics/B03676_05_01.jpg)

Go 中最常见的函数定义形式是在函数文本中包含函数的指定标识符。为了说明这一点，下表显示了几个程序的源代码，这些程序定义了具有不同参数和返回类型组合的命名函数。

<colgroup><col> <col></colgroup> 
| **代码** | **说明** |
| 

```go
package main import (
  "fmt"
  "math"
)func printPi() {
  fmt.Printf("printPi()
    %v\n", math.Pi)
} func main() {
  printPi() }               ("fmt" "math" ) func
printPi() {
  fmt.Printf("printPi()
    %v\n", math.Pi)
}
func main() { printPi() }
```

golang.fyi/ch05/func0。 | 名称标识符为[T0]的函数。它不接受任何参数，也不返回任何值。请注意，当没有要返回的内容时，`return`语句是可选的。 |
| 

```go
package main   
import "fmt"   

func avogadro() float64 {   
   return 6.02214129e23   
}   

func main() {   
   fmt.Printf("avogadro()
   = %e 1/mol\n",   
   avogadro())   
}   

```

golang.fyi/ch05/func1。 | 一个名为`avogadro.`的函数，它不接受任何参数，但返回一个类型为`float64`的值。注意，当返回值被声明为函数签名的一部分时，`return`语句是必需的。 |
| 

```go
package main   
import "fmt"    
func fib(n int) {   
  fmt.Printf("fib(%d):
    [", n)
  var p0, p1 uint64 = 0,
    1   
  fmt.Printf("%d %d ",
    p0, p1)   
  for i := 2; i <= n; i++
  {   
    p0, p1 = p1, p0+p1
    fmt.Printf("%d ",p1)
  }   
  fmt.Println("]")   
}   
func main() {   
  fib(41)   
}
```

golang.fyi/ch05/func2。 | 这定义了函数`fib`。它接受类型为`int`的参数`n`并打印最多为`n`的斐波那契序列。同样，不返回任何内容，因此省略了[T4]语句。 |
| 

```go
package main   
import (   
  "fmt"   
  "math"   
)    
func isPrime(n int) bool {   
  lim :=
  int(math.Sqrt
  (float64(n)))
  for p := 2; p <= lim;
  p++ {
    if (n % p) == 0 {   
      return false   
    }  }   
  return true   
}   
func main() {   
  prime := 37
  fmt.Printf
  ("isPrime(%d)  =
  %v\n", prime,
  isPrime(prime))
}
```

golang.fyi/ch05/func3。 | 最后一个示例定义了`isPrime` 函数。它接受一个类型为`int`的参数，并返回一个类型为`bool`的值。由于函数被声明为返回类型为`bool`的值，因此执行流中的最后一条逻辑语句必须是返回所声明类型值的`return`语句。 |

### 注

**功能签名**

一组指定的参数类型、结果类型以及这些类型的声明顺序称为函数的签名。这是帮助识别功能的另一个独特特征。两个函数可能具有相同数量的参数和结果值；但是，如果这些元素的顺序不同，则函数具有不同的签名。

## 功能类型

通常，在函数文本中声明的名称标识符用于使用调用表达式调用函数，其中函数标识符后跟参数列表。这是我们在本书中看到的，下面调用`fib` 函数的示例对此进行了说明：

```go
func main() { 
   fib(41) 
} 

```

但是，当函数的标识符不带括号出现时，它将被视为具有类型和值的正则变量，如以下程序所示：

```go
package main 
import "fmt" 

func add(op0 int, op1 int) int { 
   return op0 + op1 
} 

func sub(op0, op1 int) int { 
   return op0 - op1 
} 

func main() { 
   var opAdd func(int, int) int = add 
   opSub := sub 
   fmt.Printf("op0(12,44)=%d\n", opAdd(12, 44)) 
   fmt.Printf("sub(99,13)=%d\n", opSub(99, 13)) 
}  

```

golang.fyi/ch05/functype.go

函数的类型由其签名决定。如果函数具有相同数量的参数，且具有相同的类型，且具有相同的顺序，则函数被视为具有相同的类型。在前面的示例中，`opAdd`变量被声明为具有`func (int, int) int`类型。这是与声明的函数`add`和`sub`相同的签名。因此，`opAdd`变量被指定为`add`功能变量。这样就可以像调用`add` 函数一样调用`opAdd`。

对于`opSub`变量也是如此。它被分配由函数标识符`sub`和类型`func (int, int)`表示的值。因此，`opSub(99,13)`调用第二个函数，该函数返回减法的结果。

## 可变参数

函数的最后一个参数可以通过在参数类型前添加省略号（`…`，声明为**可变**（**可变长度参数**）。这表示在调用函数时，该类型的零个或多个值可能会传递给该函数。

以下示例实现了两个接受可变参数的函数。第一个函数计算传递值的平均值，第二个函数汇总作为参数传入的数字：

```go
package main 
import "fmt" 

func avg(nums ...float64) float64 { 
   n := len(nums) 
   t := 0.0 
   for _, v := range nums { 
         t += v 
   } 
   return t / float64(n) 
} 

func sum(nums ...float64) float64 { 
   var sum float64 
   for _, v := range nums { 
         sum += v 
   } 
   return sum 
} 

func main() { 
   fmt.Printf("avg([1, 2.5, 3.75]) =%.2f\n", avg(1, 2.5, 3.75)) 
   points := []float64{9, 4, 3.7, 7.1, 7.9, 9.2, 10} 
   fmt.Printf("sum(%v) = %.2f\n", points, sum(points...)) 
} 

```

golang.fyi/ch05/funcvariadic.go

编译器将变量参数解析为前面两个函数中类型为`[]float64`的切片。然后可以使用切片表达式访问参数值，如前一示例所示。要使用可变参数调用函数，只需提供与指定类型匹配的值的逗号分隔列表，如以下代码段所示：

```go
fmt.Printf("avg([1, 2.5, 3.75]) =%.2f\n", avg(1, 2.5, 3.75)))  

```

如果未提供任何参数，函数将接收一个空切片。精明的读者可能会想，“是否有可能将现有的一部分值作为可变参数传递？”谢天谢地，Go 提供了一个简单的习惯用法来处理这种情况。让我们在下面的代码片段中检查对`sum`函数的调用：

```go
points := []float64{9, 4, 3.7, 7.1, 7.9, 9.2, 10} 
fmt.Printf("sum(%v) = %f\n", points, sum(points...))  

```

浮点值的一部分被声明并存储在变量`points`中。通过在`sum(points...)`函数调用的参数中添加省略号，可以将切片作为可变参数传递。

## 功能结果参数

Go 函数可以定义为返回一个或多个结果值。到目前为止，在本书中，我们遇到的大多数函数都被定义为返回单个结果值。通常情况下，函数能够返回一个结果值列表，其类型多种多样，以逗号分隔（请参见上一节，*函数声明*。

为了说明这一概念，让我们检查一下下面的简单程序，它定义了一个实现欧几里得除法算法的函数（参见[http://en.wikipedia.org/wiki/Division_algorithm [T2]。`div` 函数返回商和余数值作为其结果：](http://en.wikipedia.org/wiki/Division_algorithm)

```go
package main 
import "fmt" 

func div(op0, op1 int) (int, int) { 
   r := op0 
   q := 0 
   for r >= op1 { 
         q++ 
         r = r - op1 
   } 
   return q, r 
} 

func main() { 
   q, r := div(71, 5) 
   fmt.Printf("div(71,5) -> q = %d, r = %d\n", q, r) 
} 

```

golang.fyi/ch05/funcret0.go

`**return**` 关键字后面是与函数签名中声明的结果（分别）匹配的结果值的数目。在前面的示例中，`div`函数的签名指定两个`int`值作为结果值返回。在内部，函数定义了`int`变量`p`和`r`，这些变量在函数完成时作为结果值返回。这些返回值必须与函数的签名或风险编译错误中定义的类型匹配。

必须在适当的上下文中调用具有多个结果值的函数：

*   它们必须分别分配给相同类型的标识符列表
*   它们只能包含在期望返回值数目相同的表达式中

以下源代码片段对此进行了说明：

```go
q, r := div(71, 5) 
fmt.Printf("div(71,5) -> q = %d, r = %d\n", q, r) 

```

### 命名结果参数

通常，可以使用变量标识符及其类型指定函数签名的结果列表。当使用命名标识符时，它们作为常规声明变量传递给函数，并且可以根据需要进行访问和修改。遇到`return`语句时，返回最后分配的结果值。以下源代码片段说明了这一点，它是对上一个程序的重写：

```go
func div(dvdn, dvsr int) (q, r int) { 
   r = dvdn 
   for r >= dvsr { 
         q++ 
         r = r - dvsr 
   } 
   return 
} 

```

golang.fyi/ch05/funcret1.go

注意`return`语句是裸体的；它省略了所有标识符。如前所述，`q`和`r`中分配的值将返回给调用者。为了可读性、一致性或风格，您可以选择不使用裸露的`return`语句。如前所述，在`return`语句（如`return q, r`中附加标识符的名称是完全合法的。

# 传递参数值

在 Go 中，传递给函数的所有参数都是按值传递的。这意味着在被调用函数内创建传递值的本地副本。没有通过引用传递参数值的固有概念。下面的代码通过修改`dbl` 函数中传递的参数`val`的值来说明此机制：

```go
package main 
import ( 
   "fmt" 
   "math" 
) 

func dbl(val float64) { 
   val = 2 * val // update param 
   fmt.Printf("dbl()=%.5f\n", val) 
} 

func main() { 
   p := math.Pi 
   fmt.Printf("before dbl() p = %.5f\n", p) 
   dbl(p) 
   fmt.Printf("after dbl() p = %.5f\n", p) 
} 

```

golang.fyi/ch05/funcpassbyval.go

当程序运行时，它产生以下输出，记录`p`变量在传递给`dbl`函数之前的状态。在`dbl`函数内部对传递的参数变量进行本地更新，最后调用`dbl`函数后的`p`变量值：

```go
$> go run funcpassbyval.go
before dbl() p = 3.14159
dbl()=6.28319
after dbl() p = 3.14159

```

前面的输出显示，分配给变量`p`的原始值保持变量不变，即使将其传递给似乎在内部更新其值的函数后也是如此。这是因为`dbl`函数中的`val`参数接收传递参数的本地副本。

## 实现参照传递

虽然 pass-by 值在许多情况下都是合适的，但需要注意的是，Go 可以使用指针参数值实现 pass-by-reference 语义。这允许被调用函数到达其词法范围之外，并更改指针参数所引用位置处存储的值，如以下示例中的`half`函数所做：

```go
package main 
import "fmt" 

func half(val *float64) { 
   fmt.Printf("call half(%f)\n", *val) 
   *val = *val / 2 
} 

func main() { 
   num := 2.807770 
   fmt.Printf("num=%f\n", num) 
   half(&num) 
   fmt.Printf("half(num)=%f\n", num) 
} 

```

golang.fyi/ch05/funcpassbyref.go

在前面的示例中，`main()`中对`half(&num)`函数的调用就地更新其`num`参数引用的原始值。因此，当代码执行时，它会显示`num`的原始值和调用`half`函数后的值：

```go
$> go run funcpassbyref.go
num=2.807770
call half(2.807770)
half(num)=1.403885

```

如前所述，Go 函数参数是按值传递的。即使函数将指针值作为其参数，也是如此。Go 仍然创建并传入指针值的本地副本。在前面的示例中，`half`函数通过`val`参数接收指针值的副本。代码使用指针运算符（`*`取消引用并就地操作`val`引用的值。当`half`函数退出并超出范围时，可通过调用`main` 函数访问其更改。

## 匿名函数和闭包

函数可以写为没有命名标识符的文本。这些函数称为匿名函数，可分配给稍后调用的变量，如以下示例所示：

```go
package main 
import "fmt" 

var ( 
   mul = func(op0, op1 int) int { 
         return op0 * op1 
   } 

   sqr = func(val int) int { 
         return mul(val, val) 
   } 
) 

func main() { 
   fmt.Printf("mul(25,7) = %d\n", mul(25, 7)) 
   fmt.Printf("sqr(13) = %d\n", sqr(13)) 
}  

```

golang.fyi/ch05/funcs.go

前面的程序显示了两个声明并绑定到`mul`和`sqr` 变量的匿名函数。在这两种情况下，函数都接受参数并返回值。稍后在`main()`中，变量用于调用绑定到它们的函数代码。

## 调用匿名函数文本

值得注意的是，匿名函数不必绑定到标识符。函数文本可以作为返回函数结果的表达式进行计算。这是通过以下程序中所示，在函数文本的末尾添加一系列参数值（括在括号中）来完成的：

```go
package main 
import "fmt" 

func main() { 
   fmt.Printf( 
         "94 (°F) = %.2f (°C)\n", 
         func(f float64) float64 { 
               return (f - 32.0) * (5.0 / 9.0) 
         }(94), 
   ) 
} 

```

golang.fyi/ch05/funcs.go

文字格式不仅定义匿名函数，还调用它。例如，在下面的代码段（来自上一个程序）中，匿名函数 literal 作为参数嵌套在`fmt.Printf()`中。函数本身被定义为接受一个参数并返回一个类型为`float64`的值。

```go
fmt.Printf( 
   "94 (°F) = %.2f (°C)\n", 
   func(f float64) float64 { 
         return (f - 32.0) * (5.0 / 9.0) 
   }(94), 
) 

```

由于函数文字以括号内的参数列表结尾，因此函数将作为表达式调用。

## 关闭

Go 函数文本是闭包。这意味着它们对在其封闭代码块之外声明的非局部变量具有词汇可见性。以下示例说明了这一事实：

```go
package main 
import ( 
   "fmt" 
   "math" 
) 

func main() { 
   for i := 0.0; i < 360.0; i += 45.0 { 
         rad := func() float64 { 
               return i * math.Pi / 180 
         }() 
         fmt.Printf("%.2f Deg = %.2f Rad\n", i, rad) 
   } 
} 

```

github.com/vladimirviven/learning-go/ch05/funcs.go

在前面的程序中，函数文字代码块`func() float64 {return deg * math.Pi / 180}()`被定义为一个将度转换为弧度的表达式。在循环的每次迭代中，封闭函数文字和外部非局部变量`i`之间形成一个闭包。这提供了一种更简单的习惯用法，其中函数自然地访问非本地值，而无需求助于指针等其他方法。

### 注

在 Go 中，在创建闭包的外部函数超出范围很久之后，词汇闭包值可以保持其闭包的边界。垃圾收集器将在这些闭合值变得无界时处理清理。

# 高阶函数

我们已经确定 Go 函数是绑定到类型的值。因此，Go 函数可以将另一个函数作为参数，也可以将一个函数作为结果值返回，这并不奇怪。这描述了被称为高阶函数的概念，这是一个从数学中采用的概念。[T0]等类型让程序员抽象数据，而高阶函数提供了一种封装和抽象行为的机制，这些行为可以组合在一起形成更复杂的行为。

为了让这个概念更清楚，让我们来看看下面的程序，它使用一个高阶函数[T0]，来做三件事。它接受整数切片和函数作为参数。它将指定的函数应用于切片中的每个元素。最后，`apply`函数还返回一个函数作为其结果：

```go
package main 
import "fmt" 

func apply(nums []int, f func(int) int) func() { 
   for i, v := range nums { 
         nums[i] = f(v) 
   } 
   return func() { 
         fmt.Println(nums) 
   } 
} 

func main() { 
   nums := []int{4, 32, 11, 77, 556, 3, 19, 88, 422} 
   result := apply(nums, func(i int) int { 
         return i / 2 
   }) 
   result() 
} 

```

golang.fyi/ch05/funchighorder.go

在程序中，`apply`函数由一个匿名函数调用，该函数将切片中的每个元素减半，如以下代码段所示：

```go
nums := []int{4, 32, 11, 77, 556, 3, 19, 88, 422} 
result := apply(nums, func(i int) int { 
   return i / 2 
}) 
result() 

```

作为一个高阶函数，`apply`抽象了可由`func(i int) int`类型的任何函数提供的转换逻辑，如下所示。由于`apply`函数返回一个函数，因此可以调用变量`result`，如前一段所示。

在探索本书和 Go 语言时，您将继续遇到高阶函数的用法。这是一个流行的习惯用法，在标准库中大量使用。您还可以在一些并发模式中找到用于分配工作负载的高阶函数（请参见[第 9 章](09.html "Chapter 9. Concurrency")、*并发*。

# 错误信号及处理

现在，让我们讨论一下在进行函数调用时如何惯用地发出错误信号并处理错误。如果您使用过 Python、Java 或 C#等语言，您可能熟悉在出现不希望出现的状态时通过抛出异常来中断执行代码的流程。

正如我们将在本节中探讨的那样，Go 有一种简化的错误信号和错误处理方法，它使程序员有责任在调用的函数返回后立即处理可能的错误。Go 不鼓励通过不加区别地用异常短接正在执行的程序来中断执行，希望它能在调用堆栈的更上层得到正确处理。在 Go 中，发送错误信号的传统方式是在函数执行过程中出现问题时返回类型为`error`的值。因此，让我们仔细看看这是如何做到的。

## 信令错误

为了更好地理解上一段中描述的内容，让我们从一个示例开始。以下源代码实现了一个字谜程序，如乔恩·本特利（Jon Bentley）的《编程珍珠》（第二版）一书中第 2 栏所述。代码读取一个字典文件（`dict.txt`），并使用相同的字谜对所有单词进行分组。如果代码不太合理，请参见[golang.fyi/ch05/anagram1.go](http://learning.golang.fyi/ch05/anagram1.go)了解程序各部分工作原理的注释说明。

```go
package main 

import ( 
   "bufio" 
   "bytes" 
   "fmt" 
   "os" 
   "errors" 
) 

// sorts letters in a word (i.e. "morning" -> "gimnnor") 
func sortRunes(str string) string { 
   runes := bytes.Runes([]byte(str)) 
   var temp rune 
   for i := 0; i < len(runes); i++ { 
         for j := i + 1; j < len(runes); j++ { 
               if runes[j] < runes[i] { 
                     temp = runes[i] 
                     runes[i], runes[j] = runes[j], temp 
               } 

         } 
   } 
   return string(runes) 
} 

// load loads content of file fname into memory as []string 
func load(fname string) ([]string, error) { 
   if fname == "" { 
         return nil, errors.New( 
               "Dictionary file name cannot be empty.")  
   } 

   file, err := os.Open(fname) 
   if err != nil { 
         return nil, err 
   } 
   defer file.Close() 

   var lines []string 
   scanner := bufio.NewScanner(file) 
   scanner.Split(bufio.ScanLines) 
   for scanner.Scan() { 
         lines = append(lines, scanner.Text()) 
   } 
   return lines, scanner.Err() 
} 

func main() { 
   words, err := load("dict.txt")       
   if err != nil { 
         fmt.Println("Unable to load file:", err) 
         os.Exit(1) 
   } 

      anagrams := make(map[string][]string) 
   for _, word := range words { 
         wordSig := sortRunes(word) 
         anagrams[wordSig] = append(anagrams[wordSig], word) 
   } 

   for k, v := range anagrams { 
         fmt.Println(k, "->", v) 
   } 
} 

```

golang.fyiy/ch05/anagram1。

同样，如果您想要对上一个程序进行更详细的解释，请查看前面提供的链接。这里的重点是前一个程序中使用的错误信号。按照惯例，Go 代码使用内置类型`error`在函数执行期间发生错误时发出信号。因此，函数必须返回 error 类型的值，以向其调用者指示出现了问题。以下是`load`函数的片段（摘自上一个示例）：

```go
func load(fname string) ([]string, error) { 
   if fname == "" { 
       return nil, errors.New( 
         "Dictionary file name cannot be empty.")  
   } 

   file, err := os.Open(fname) 
   if err != nil { 
         return nil, err 
   } 
   ... 
} 

```

请注意，`load`函数返回多个结果参数。一个是期望值，在本例中为`[]string`，另一个是错误值。惯用的 Go 表示程序员为类型为`error`的结果返回一个非 nil 值，以指示在函数执行过程中发生了异常。在前面的代码片段中，`load`函数在两种可能的情况下向其调用者发出错误发生的信号：

*   当预期文件名（`fname`为空时
*   调用`os.Open()`失败时（例如权限错误或其他）

在第一种情况下，当没有提供文件名时，代码返回一个错误，使用`errors.New()`创建一个`error`类型的值以退出函数。在第二种情况下，`os.Open`函数返回一个指针，表示文件和分别分配给`file`和`err` 变量的错误。如果`err`不是`nil`（意味着生成了错误），`load`函数的执行会提前停止，`err`的值会返回给调用函数在调用堆栈的更上层进行处理。

### 注

当为具有多个结果参数的函数返回错误时，通常为其他（非错误类型）参数返回零值。在该示例中，为类型为`[]string`的结果返回一个值`nil`。虽然没有必要，但它简化了错误处理，避免了函数调用方的任何混淆。

## 错误处理

如前所述，错误状态的信令与在函数执行期间返回类型为`error`的非零值一样简单。调用方可以选择处理`error`或`return`它，以便像在`load`函数中所做的那样对调用堆栈进行进一步评估。这个习惯用法迫使错误向上传播，直到在某个时候被处理为止。下一个代码片段显示了在`main`函数中如何处理`load`函数生成的错误：

```go
func main() { 
   words, err := load("dict.txt") 
   if err != nil { 
         fmt.Println("Unable to load file:", err) 
         os.Exit(1) 
   } 
   ... 
} 

```

由于`main`函数是调用堆栈中最顶层的调用方，因此它通过终止整个程序来处理错误。

这就是围棋中错误处理机制的全部内容。该语言强制程序员在返回类型为[T0]的值的每个函数调用上始终测试错误状态。对于某些人来说，`if…not…nil error`处理习惯用法可能显得过于冗长，尤其是如果您来自一种具有正式异常机制的语言。然而，这里的好处是，程序可以构造一个健壮的执行流，程序员总是知道错误可能来自哪里，并适当地处理它们。

## 错误类型

`error`类型是一个内置接口，因此必须先实现才能使用。幸运的是，Go 标准库附带了可供使用的实现。我们已经使用了包[T1]中的一个实现：

```go
errors.New("Dictionary file name cannot be empty.")  

```

您还可以使用`fmt.Errorf`函数创建参数化错误值，如以下代码段所示：

```go
func load(fname string) ([]string, error) { 
   if fname == "" { 
         return nil, errors.New( 
             "Dictionary file name cannot be emtpy.") 
   } 

   file, err := os.Open(fname) 
   if err != nil { 
         return nil, fmt.Errorf( 
             "Unable to open file %s: %s", fname, err) 
   } 
   ... 
} 

```

golang.fyi/ch05/anagram2.go

将错误值分配给高级变量也是惯用做法，这样它们就可以根据需要在整个程序中重用。以下代码片段取自[http://golang.org/src/os/error.go](http://golang.org/src/os/error.go) 显示与操作系统文件操作相关的可重用错误声明：

```go
var ( 
   ErrInvalid    = errors.New("invalid argument") 
   ErrPermission = errors.New("permission denied") 
   ErrExist      = errors.New("file already exists") 
   ErrNotExist   = errors.New("file does not exist") 
) 

```

T01230551；http://golang.org/src/os/error.go 1.123051.

您还可以创建自己的`error`接口实现来创建自定义错误。本主题在[第 7 章](08.html "Chapter 8. Methods, Interfaces, and Objects")中*方法、接口和对象*中再次讨论，本书在其中讨论了扩展类型的概念。

# 推迟函数调用

Go 支持延迟函数调用的概念。在函数调用之前放置关键字`defer`有一个有趣的效果，即将函数推到内部堆栈，在封闭函数返回之前延迟其执行。为了更好地解释这一点，让我们从以下简单的程序开始，该程序演示了[T1]的用法：

```go
package main 
import "fmt" 

func do(steps ...string) { 
   defer fmt.Println("All done!") 
   for _, s := range steps { 
         defer fmt.Println(s) 
   } 

   fmt.Println("Starting") 
} 

func main() { 
   do( 
         "Find key", 
         "Aplly break", 
         "Put key in ignition", 
         "Start car", 
   ) 
} 

```

golang.fyi/ch05/1.go

上例定义了采用可变参数`steps`的`do`函数。函数使用[T2]延迟语句。接下来，函数循环通过切片`steps`并使用`defer fmt.Println(s).`延迟每个元素的输出。函数`do`中的最后一条语句是一个非延迟调用，以在程序执行时通知`fmt.Println("Starting").`打印字符串值的顺序，如下输出所示：

```go
$> go run defer1.go
Starting
Start car
Put key in ignition
Aplly break
Find key
All done!

```

有几个事实可以解释打印输出的相反顺序。首先，回想一下，延迟函数是在其封闭函数返回之前执行的。因此，打印的第一个值是由最后一个非延迟方法调用生成的。接下来，如前所述，将延迟语句推送到堆栈中。因此，使用后进先出顺序执行延迟调用。这就是为什么`"All done!"`是输出中打印的最后一个字符串值。

## 使用延时

`defer`关键字通过延迟函数调用来修改程序的执行流。此功能的一个惯用用法是进行资源清理。由于当周围函数返回时，DEBER 将始终执行，因此它是附加清理代码的好地方，例如：

*   关闭打开的文件
*   释放网络资源
*   关闭围棋频道
*   提交数据库事务
*   继续

为了说明这一点，让我们回到前面的字谜示例。下面的代码段显示了一个版本的代码，其中在加载文件后，使用 defer 关闭文件。`load`函数在返回之前调用`file.Close()`：

```go
func load(fname string) ([]string, error) { 
... 
   file, err := os.Open(fname) 
   if err != nil { 
         return nil, err 
   } 
   defer file.Close() 
... 
} 

```

golang.fyi/ch05/anagram2.go

开放-延迟-关闭资源的模式在 Go 中被广泛使用。通过在打开或创建资源后立即放置延迟意图，允许代码自然读取，并降低创建资源泄漏的可能性。

# 功能恐慌与恢复

本章前面提到，Go 没有其他语言提供的传统异常机制。然而，在 Go 中，有一种方法可以突然退出正在执行的函数，称为函数恐慌。相反，当程序陷入恐慌时，Go 提供了一种恢复和重新获得对执行流控制的方法。

## 功能性恐慌

在执行过程中，函数可能会因以下任一原因而死机：

*   显式调用**恐慌**内置函数
*   使用由于异常状态而崩溃的源代码包
*   访问 nil 值或越界数组元素
*   并发死锁

当函数崩溃时，它会中止并执行其延迟的调用。然后它的调用者惊慌失措，引起连锁反应，如下图所示：

![Function panic](graphics/B03676_05_02.jpg)

紧急顺序一直持续到调用堆栈，直到到达`main`函数且程序退出（崩溃）。下面的源代码片段显示了一个版本的 anagram 程序，如果输出 anagram 文件在尝试创建时已经存在，该程序将导致显式死机。这样做是为了在出现文件错误时导致`write`函数死机：

```go
package main 
... 
func write(fname string, anagrams map[string][]string) { 
   file, err := os.OpenFile( 
         fname,  
         os.O_WRONLY+os.O_CREATE+os.O_EXCL,  
         0644, 
   ) 
   if err != nil { 
         msg := fmt.Sprintf( 
               "Unable to create output file: %v", err, 
         ) 
         panic(msg) 
   } 
   ... 
} 

func main() { 
   words, err := load("dict.txt") 
   if err != nil { 
         fmt.Println("Unable to load file:", err) 
         os.Exit(1) 
   } 
   anagrams := mapWords(words) 
   write("out.txt", anagrams) 
} 

```

golang.fyi/ch05/anagram2.go

在前面的代码段中，`write`函数在`os.OpenFile()`方法出错时调用`panic`函数。当程序调用`main`函数时，如果工作目录中已经有输出文件，程序将死机，如下堆栈跟踪所示，指示导致死机的调用顺序：

```go
> go run anagram2.go 
panic: Unable to create output file: open out.txt: file exists
goroutine 1 [running]:
main.write(0x4e7b30, 0x7, 0xc2080382a0)
/Go/src/github.com/vladimirvivien/learning-go/ch05/anagram2.go:72 +0x1a3 
main.main()
Go/src/github.com/vladimirvivien/learning-go/ch05/anagram2.go:103 +0x1e9
exit status 2

```

## 功能性恐慌恢复

如前所述，当函数崩溃时，它会使整个程序崩溃。这可能是期望的结果，具体取决于您的需求。然而，在恐慌序列开始后，有可能重新获得控制。为此，Go 提供了名为`recover`的内置函数。

恢复工作与恐慌同步进行。调用函数 recover 返回作为参数传递给 panic 的值。下面的代码显示了如何从上一个示例中引入的紧急调用中恢复。在这个版本中，为了清晰起见，写入功能被移动到`makeAnagram()`内部。当`write`函数从`makeAnagram()` 调用，无法打开文件时，会死机。但是，现在添加了其他代码来恢复：

```go
package main 
... 
func write(fname string, anagrams map[string][]string) { 
   file, err := os.OpenFile( 
         fname,  
         os.O_WRONLY+os.O_CREATE+os.O_EXCL,  
         0644, 
   ) 
   if err != nil { 
         msg := fmt.Sprintf( 
               "Unable to create output file: %v", err, 
         ) 
         panic(msg) 
   } 
   ... 
} 

func makeAnagrams(words []string, fname string) { 
   defer func() { 
         if r := recover(); r != nil { 
               fmt.Println("Failed to make anagram:", r) 
         } 
   }() 

   anagrams := mapWords(words) 
   write(fname, anagrams) 
} 
func main() { 
   words, err := load("") 
   if err != nil { 
         fmt.Println("Unable to load file:", err) 
         os.Exit(1) 
   } 
   makeAnagrams(words, "") 
} 

```

golang.fyi/ch05/anagram3.go

为了能够从展开的死机序列中恢复，代码必须对 recover 函数进行延迟调用。在前面的代码中，这是在`makeAnagrams`函数中完成的，方法是将`recover()`包装在匿名函数文本中，如以下代码段中所强调的：

```go
defer func() { 
   if r := recover(); r != nil { 
         fmt.Println("Failed to make anagram:", r) 
   } 
}() 

```

当执行延迟`recover`功能时，程序有机会重新获得控制，并防止死机破坏正在运行的程序。如果`recover()`返回`nil`，则表示当前没有向上展开调用堆栈的紧急情况，或者紧急情况已经在下游处理。

因此，现在当程序执行时，程序将恢复并优雅地显示问题，而不是因堆栈跟踪而崩溃，如以下输出所示：

```go
> go run anagram3.go
Failed to make anagram: Unable to open output file for creation: open out.txt: file exists

```

### 注

您可能想知道，当在对`panic`的调用中传递字符串时，为什么我们要使用`nil`来测试`recover`函数返回的值。这是因为 panic 和 recover 都采用空接口类型。您将了解到，空接口类型是一种泛型类型，能够表示 Go 类型系统中的任何类型。我们将在[第 7 章](08.html "Chapter 8. Methods, Interfaces, and Objects")、*方法、接口和对象*中对空接口进行详细介绍。

# 总结

本章向读者介绍了 Go 函数的探索。首先概述命名函数声明，然后讨论函数参数。本章深入讨论了函数类型和函数值。本章的最后一部分讨论了错误处理、恐慌和恢复的语义。下一章继续讨论功能；但是，它是在 Go 包的上下文中实现的。它解释了包作为 Go 函数（和其他代码元素）的逻辑分组的作用，以形成可共享和可调用的代码模块。